cpp/src/arrow/array/util.cc

```diff
@@ -296,11 +301,14 @@ class ArrayDataEndianSwapper {
 namespace internal {
 
 Result<std::shared_ptr<ArrayData>> SwapEndianArrayData(
-    const std::shared_ptr<ArrayData>& data) {
+    const std::shared_ptr<ArrayData>& data, MemoryPool* pool) {
   if (data->offset != 0) {
     return Status::Invalid("Unsupported data format: data.offset != 0");
   }
-  ArrayDataEndianSwapper swapper(data);
+  if (pool == nullptr) {
+    return Status::Invalid("Unsupported pool == nullptr");
+  }
```

Comment:
This being an internal API, should we even have this validation?

Reply:
No indeed!

Reply:
Also, in general it's assumed that people don't pass nullptr where a pointer is expected. There may be specific exceptions where the argument is optional.

Reply:
Removed now

Reply:
I think a `DCHECK(pool)` is enough here.

The reason we pass `*` instead of `&` in functions like this is that we follow the Google Style convention that says non-const "reference" parameters should be pointers instead of `&`. A consequence of that convention is that we expect the caller to do the right thing and pass a non-null mutable parameter unless stated in the function documentation that `nullptr` is allowed.

Reply:
( I only saw your comments after I posted mine) :)